egui/src/containers/collapsing_header.rs

```diff
@@ -3,102 +3,256 @@ use std::hash::Hash;
-            Some(true)
-        } else {
-            State::load(ctx, id).map(|state| state.open)
-        }
+    pub fn is_open(&self) -> bool {
+        self.state.open
+    }
+
+    pub fn set_open(&mut self, open: bool) {
+        self.state.open = open;
     }
 
     pub fn toggle(&mut self, ui: &Ui) {
-        self.open = !self.open;
+        self.state.open = !self.state.open;
         ui.ctx().request_repaint();
     }
 
     /// 0 for closed, 1 for open, with tweening
-    pub fn openness(&self, ctx: &Context, id: Id) -> f32 {
+    pub fn openness(&self, ctx: &Context) -> f32 {
         if ctx.memory().everything_is_visible() {
             1.0
         } else {
-            ctx.animate_bool(id, self.open)
+            ctx.animate_bool(self.id, self.state.open)
+        }
+    }
+
+    /// Will toggle when clicked, etc.
+    pub(crate) fn show_default_button_with_size(
+        &mut self,
+        ui: &mut Ui,
+        button_size: Vec2,
+    ) -> Response {
+        let (_id, rect) = ui.allocate_space(button_size);
+        let response = ui.interact(rect, self.id, Sense::click());
+        if response.clicked() {
+            self.toggle(ui);
+        }
+        let openness = self.openness(ui.ctx());
+        paint_default_icon(ui, openness, &response);
+        response
+    }
+
+    /// Will toggle when clicked, etc.
+    fn show_default_button_indented(&mut self, ui: &mut Ui) -> Response {
+        let size = Vec2::new(ui.spacing().indent, ui.spacing().icon_width);
+        let (_id, rect) = ui.allocate_space(size);
+        let response = ui.interact(rect, self.id, Sense::click());
+        if response.clicked() {
+            self.toggle(ui);
+        }
+
+        let (mut icon_rect, _) = ui.spacing().icon_rectangles(response.rect);
+        icon_rect.set_center(pos2(
+            response.rect.left() + ui.spacing().indent / 2.0,
+            response.rect.center().y,
+        ));
+        let openness = self.openness(ui.ctx());
+        let small_icon_response = Response {
+            rect: icon_rect,
+            ..response.clone()
+        };
+        paint_default_icon(ui, openness, &small_icon_response);
+        response
+    }
+
+    /// Shows header and body (if expanded).
+    ///
+    /// The header will start with the default button in a horizontal layout, followed by whatever you add.
+    ///
+    /// Will also store the state.
+    ///
+    /// Returns the response of the collapsing button, the custom header, and the custom body.
+    ///
+    /// ```
+    /// # egui::__run_test_ui(|ui| {
+    /// let id = ui.make_persistent_id("my_collapsing_header");
+    /// egui::collapsing_header::CollapsingState::load_with_default_open(ui.ctx(), id, false)
+    ///     .show_custom_header(ui, |ui| {
+    ///         ui.label("Header"); // you can put checkboxes or whatever here
+    ///     })
+    ///     .body(|ui| ui.label("Body"));
+    /// # });
+    /// ```
+    pub fn show_custom_header<HeaderRet>(
```

Comment:
Why not `header` or `show_header`? It's looks that header always "custom", so not need to explicitly define that?

Reply:
`egui` uses the `show` prefix for all container widgets that take an `impl FnOnce(&mut Ui)` (`Window::show`, `Frame::show`, …).

…but I agree with `show_header` being shorter and sweeter